keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,720 @@
+comment in [kubernetes #82125] has more details on the history of how
+we got to where we are now, or at least to where we were before
+dockershim was removed.)
+
+We should clean this up so that kubelet no longer creates any IPTables
+chains, and that kube-proxy creates all of the IPTables chains it
+needs.
+
+[the removal of dockershim]: ../../sig-node/2221-remove-dockershim
+[kubernetes #82125]: https://github.com/kubernetes/kubernetes/issues/82125
+
+## Motivation
+
+### Goals
+
+- Rewrite kube-proxy's packet processing to no longer use the
+  `KUBE-MARK-DROP` chain created by kubelet.
+
+    - Ensure that the proxy implementations in kpng get updated
+      accordingly as well.
+
+- Remove most IPTables chain management from kubelet.
+
+    - Remove the `--iptables-drop-bit` and `--iptables-masquerade-bit`
+      arguments from kubelet.
+
+    - Ensure that users and third-party software that make assumptions
+      about kubelet's IPTables rules have time to get updated for the
+      change in rule creation.
+
+    - Kubelet will continue to create at least one IPTables chain as a
+      hint to `iptables-wrapper` and other components that need to
+      know whether the system is using `legacy` or `nft` iptables.
+
+- Document for the future that Kubernetes's IPTables chains (other
+  than the IPTables mode "hint" chain) are meant for its internal use
+  only, do not constitute API, and should not be relied on by external
+  components.
+
+### Non-Goals
+
+- Changing `KUBE-MARK-MASQ` and `KUBE-MARK-DROP` to use the connection
+  mark or connection label rather than the packet label (as discussed
+  in [kubernetes #78948]).
+
+[kubernetes #78948]: https://github.com/kubernetes/kubernetes/issues/78948
+
+```
+<<[UNRESOLVED other-kubelet-network-cruft ]>>
+
+Is there any other related legacy networking cruft in kubelet that it
+would make sense to remove/move at the same time as the IPTables
+stuff?
+
+<<[/UNRESOLVED]>>
+```
+
+## Proposal
+
+### Notes/Constraints/Caveats
+
+#### The Chains in Question and Their Purpose
+
+Kubelet currently creates five IPTables chains; two to help with
+masquerading packets, two to help with dropping packets, and one for
+purely-internal purposes:
+
+  - `KUBE-MARK-MASQ` and `KUBE-POSTROUTING`
+
+    `KUBE-MARK-MASQ` marks packets as needing to be masqueraded (by
+    setting a configurable bit of the "packet mark").
+    `KUBE-POSTROUTING` checks the packet mark and calls `-j
+    MASQUERADE` on the packets that were previously marked for
+    masquerading.
+
+    These chains were formerly used for HostPort handling in
+    dockershim, but are no longer used by kubelet. Kube-proxy (in
+    iptables or ipvs mode) creates identical copies of both of these
+    chains, which it uses for service handling.
+
+  - `KUBE-MARK-DROP` and `KUBE-FIREWALL`
+
+    `KUBE-MARK-DROP` marks packets as needing to be dropped by setting
+    a _different_ configurable bit on the packet mark. `KUBE-FIREWALL`
```

Comment:
KUBE-FIREWALL was also extended to deal with the route_localnet CVE
https://github.com/kubernetes/kubernetes/pull/91569

Reply:
Yeah, forgot about that.

So this is its own mess: 91569 only talks about NodePort, but the problem theoretically applies to HostPorts too. Except, it only applies if you set `route_localnet`, which kube-proxy does and kubelet doesn't... so for our own purposes we can move 91569's rule from kubelet to kube-proxy. (In fact, it should have been there originally, but `KUBE-FIREWALL` sounds like a good place to put it, and that's created from kubelet, so it got put there.)

But given that it was previously in kubelet, there are probably alternate proxies and hostport implementations that would be vulnerable if we didn't have this rule... so we may have to keep creating it...

We should probably have a KEP about eventually getting rid of localhost NodePort/HostPort behavior...

Reply:
I have this printed in my table 
![image](https://user-images.githubusercontent.com/6450081/155537288-bc8c8896-ef9f-469f-9069-47e3676b828f.png)